### PR TITLE
[libc++] P2389R2: `dextents` Index Type Parameter

### PR DESCRIPTION
**Pull request by xiaoyang-sde**
## Introduction

This patch implements [P2389R2](https://wg21.link/P2389R2), which was adopted at the St. Louis meeting. It builds upon previous enhancements from [P2299R3](https://wg21.link/P2299R3), which introduced deduction guides and the `dextents` alias template. The main problem addressed is the increased verbosity of `dextents` following the parameterization of index types in [P2553R1](https://wg21.link/P2553R1). After considering and rejecting a breaking change, the paper proposes introducing a new template alias called `dims`, which simplifies the declaration of `mdspan` with dynamic extents while maintaining the option to customize the index type.

## Reference

- [P2389R2](https://wg21.link/P2389R2)
- [Draft C++ Standard: [mdspan.syn]](https://eel.is/c++draft/mdspan.syn)
- [Draft C++ Standard: [version.syn]](https://eel.is/c++draft/version.syn)

### CODE REVIEW

File: libcxx/docs/ReleaseNotes/19.rst [hunk 1]
```diff
@@ -53,6 +53,7 @@ Implemented Papers
 - P2713R1 - Escaping improvements in ``std::format``
 - P2231R1 - Missing ``constexpr`` in ``std::optional`` and ``std::variant``
 - P0019R8 - ``std::atomic_ref``
+- P2389R2 - ``dextents`` Index Type Parameter
```
Comment (dalg24) [inline libcxx/docs/ReleaseNotes/19.rst:56]: Should we depart from the paper name to say "alias template dims for the extents of a mdspan" or such?
I don't know what is the common practice but I find the actual paper title not so descriptive.

File: libcxx/include/__mdspan/extents.h [hunk 2]
```diff
@@ -465,6 +465,12 @@ template <class... _IndexTypes>
 explicit extents(_IndexTypes...) -> extents<size_t, size_t(((void)sizeof(_IndexTypes), dynamic_extent))...>;
 #  endif
 
+#  if _LIBCPP_STD_VER >= 26
+// [mdspan.extents.dims], alias template `dims`
+template <size_t _Rank, class _IndexType = size_t>
+using dims = dextents<_IndexType, _Rank>;
+#  endif
```
Comment (dalg24) [inline libcxx/include/__mdspan/extents.h:472]: I would prefer if we moved it right after the definition of `dextents`, that is L457.

**Review by dalg24 — APPROVED** [review]
Looks good, I would prefer if you moved the definition of the `dims` alias template right after the definition of `dextents`.

File: libcxx/docs/ReleaseNotes/19.rst (hunk 1 above)
Comment (xiaoyang-sde) [reply libcxx/docs/ReleaseNotes/19.rst:56]: Thank you for the review. I will update the code once this comment has been addressed.

**Review by ldionne — APPROVED** [review]
LGTM w/ @dalg24 's comments addressed.

File: libcxx/docs/ReleaseNotes/19.rst (hunk 1 above)
Comment (xiaoyang-sde) [reply libcxx/docs/ReleaseNotes/19.rst:56]: @ldionne Would you mind share your thoughts on this?

Comment (ldionne) [reply libcxx/docs/ReleaseNotes/19.rst:56]: In general I would stick to the paper title, but in this case the paper title is.. not really helpful. I would agree with @dalg24 's suggestion.

**Comment by mordante** [comment]
I've just committed https://github.com/llvm/llvm-project/pull/97951 can you rebase your patch on main and update the status page for this patch.

**Comment by ldionne** [comment]
Rebased and updated the status page. This can be merged once the CI is green.

**Pull request merged**